### PR TITLE
Check Contribution Rule

### PR DESCRIPTION
**Pull request by muthusubra**
## What

[Link to story](https://dsdmoj.atlassian.net/browse/LCAM-XXX)

Describe what you did and why.

## Checklist

Before you ask people to review this PR:

- [ ] Tests should be passing: `./gradlew test`
- [ ] Github should not be reporting conflicts; you should have recently run `git rebase main`.
- [ ] Avoid mixing whitespace changes with code changes in the same commit. These make diffs harder to read and conflicts more likely.
- [ ] You should have looked at the diff against main and ensured that nothing unexpected is included in your changes.
- [ ] You should have checked that the commit messages say why the change was made.

### CODE REVIEW

**Review by venkat980 — APPROVED** [review]

File: crown-court-contribution/src/main/java/uk/gov/justice/laa/crime/contribution/controller/ContributionController.java [hunk 1]
```diff
@@ -88,4 +92,27 @@ public ResponseEntity<Void> requestTransfer(@Valid @RequestBody ApiContributionT
...
+    )
+    @DefaultHTTPErrorResponse
+    public ResponseEntity<Boolean> checkContributionRule(
+            @Parameter(description = "Data required to check contribution rule",
+                    content = @Content(mediaType = MediaType.APPLICATION_JSON_VALUE,
+                            schema = @Schema(implementation = ApiCalculateContributionRequest.class)
+                    )
+            )
+            @Valid @RequestBody
+            ApiMaatCalculateContributionRequest maatCalculateContributionRequest) {
```
Comment (Mhowell494) [inline crown-court-contribution/src/main/java/uk/gov/justice/laa/crime/contribution/controller/ContributionController.java:110]: There are quite a few mandatory fields in this request object, such as the `RepId`, `applicationId` etc. that I don't really think we should be passing? So imo we'd be better off just creating a new schema for this endpoint.

File: crown-court-contribution/src/test/java/uk/gov/justice/laa/crime/contribution/controller/ContributionControllerContributionIntegrationTest.java [hunk 2]
```diff
@@ -47,7 +49,10 @@ class ContributionControllerContributionIntegrationTest {
     private MockMvc mvc;
     private static final WireMockServer wiremock = new WireMockServer(9999);
     private static final String ENDPOINT_URL = "/api/internal/v1/contribution/calculate-contribution";
+    private static final String BASE_URL = "/api/internal/v1/contribution/";
```
Comment (Mhowell494) [inline crown-court-contribution/src/test/java/uk/gov/justice/laa/crime/contribution/controller/ContributionControllerContributionIntegrationTest.java:52]: Is this going to cause problems? It looks like there will be a double slash.

File: crown-court-contribution/src/test/java/uk/gov/justice/laa/crime/contribution/controller/ContributionControllerTest.java [hunk 3]
```diff
@@ -41,6 +44,7 @@ class ContributionControllerTest {
 
     private static final String BASE_URL = "/api/internal/v1/contribution/";
     private static final String ENDPOINT_URL = BASE_URL + "calculate-contribution";
+    private static final String CHECK_CONTRIBUTION_RULE_ENDPOINT_URL = BASE_URL + "/check-contribution-rule";
```
Comment (Mhowell494) [inline crown-court-contribution/src/test/java/uk/gov/justice/laa/crime/contribution/controller/ContributionControllerTest.java:47]: Potentially same issue with the double slash in the final URL.

**Review by Mhowell494 — CHANGES_REQUESTED** [review]

File: crown-court-contribution/src/test/java/uk/gov/justice/laa/crime/contribution/controller/ContributionControllerContributionIntegrationTest.java (hunk 2 above)
Comment (muthusubra) [reply crown-court-contribution/src/test/java/uk/gov/justice/laa/crime/contribution/controller/ContributionControllerContributionIntegrationTest.java]: Fixed it

File: crown-court-contribution/src/test/java/uk/gov/justice/laa/crime/contribution/controller/ContributionControllerTest.java
Comment (Mhowell494) [inline crown-court-contribution/src/test/java/uk/gov/justice/laa/crime/contribution/controller/ContributionControllerTest.java:1]: Need to update the `givenValidRequest_whenCheckContributionRuleIsInvoked_thenOkResponse` test, it looks like we're still passing in the old schema. I'd say we also need at least one more. test to check throwing a badRequest if the input validation fails.

File: crown-court-contribution/src/test/java/uk/gov/justice/laa/crime/contribution/controller/ContributionControllerContributionIntegrationTest.java [hunk 4]
```diff
@@ -293,10 +292,10 @@ void givenApiMaatCalculateContributionRequest_whenCheckContributionRuleIsInvoked
...
-        ApiMaatCalculateContributionRequest calculateContributionRequest =
-                TestModelDataBuilder.buildCalculateContributionRequest();
-        calculateContributionRequest.setCrownCourtOutcome(List.of(getApiCrownCourtOutcome(CrownCourtOutcome.AQUITTED)));
-        String requestData = objectMapper.writeValueAsString(calculateContributionRequest);
+        ApiMaatCheckContributionRuleRequest apiMaatCheckContributionRuleRequest =
+                TestModelDataBuilder.buildCheckContributionRuleRequest();
+        apiMaatCheckContributionRuleRequest.setCrownCourtOutcome(List.of(getApiCrownCourtOutcome(CrownCourtOutcome.AQUITTED)));
+        String requestData = objectMapper.writeValueAsString(apiMaatCheckContributionRuleRequest);
         mvc.perform(buildRequestGivenContent(HttpMethod.POST, requestData, CHECK_CONTRIBUTION_RULE_ENDPOINT_URL))
                 .andExpect(status().isOk())
```
Comment (Mhowell494) [inline crown-court-contribution/src/test/java/uk/gov/justice/laa/crime/contribution/controller/ContributionControllerContributionIntegrationTest.java:300]: I think we should also check the boolean response object here too.

File: crown-court-contribution/src/main/resources/schemas/apiMaatCheckContributionRuleRequest.json [hunk 5]
```diff
@@ -0,0 +1,28 @@
+{
+  "$schema": "http://json-schema.org/draft-04/schema#",
+  "id": "apiMaatCheckContributionRuleRequest.json",
+  "type": "object",
+  "title": "Calculate Contribution Request from MAAT application",
```
Comment (Mhowell494) [inline crown-court-contribution/src/main/resources/schemas/apiMaatCheckContributionRuleRequest.json:5]: Need to update the title and description.

**Review by Mhowell494 — CHANGES_REQUESTED** [review]

File: crown-court-contribution/src/main/resources/schemas/apiMaatCheckContributionRuleRequest.json (hunk 5 above)
Comment (muthusubra) [reply crown-court-contribution/src/main/resources/schemas/apiMaatCheckContributionRuleRequest.json:5]: Updated

File: crown-court-contribution/src/test/java/uk/gov/justice/laa/crime/contribution/controller/ContributionControllerContributionIntegrationTest.java (hunk 4 above)
Comment (muthusubra) [reply crown-court-contribution/src/test/java/uk/gov/justice/laa/crime/contribution/controller/ContributionControllerContributionIntegrationTest.java:300]: Already checking the boolean response at line 302

File: crown-court-contribution/src/test/java/uk/gov/justice/laa/crime/contribution/controller/ContributionControllerTest.java
Comment (muthusubra) [reply crown-court-contribution/src/test/java/uk/gov/justice/laa/crime/contribution/controller/ContributionControllerTest.java:1]: Updated the test and also added aa test for bad request

**Review by Mhowell494 — APPROVED** [review]

**Pull request merged**